### ISSUE TITLE
jlibtorrent - SessionManager Fails to Bind to listenInterfaces

### ISSUE THREAD

**Issue by LakshmanSagar** [frostwire/frostwire-jlibtorrent#302]
### Description

`jlibtorrent`'s `SessionManager` fails to bind to the specified `listenInterfaces`, falling back to `0.0.0.0:0`, resulting in binding failures with the following error:

```
onListenFailed(): iface= , address= 0.0.0.0, port= 0, socketType= TCP, errorCode= com.frostwire.jlibtorrent.ErrorCode
error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
```

Despite applying a valid configuration (e.g., `0.0.0.0:6881-6889,[::]:6881-6889`) through `SettingsPack`, the settings are either overridden or ignored by the `SessionManager`.

---

### Steps to Reproduce

1. Configure `SettingsPack` as follows:
    ```kotlin
    settingsPack.listenInterfaces("0.0.0.0:6881-6889,[::]:6881-6889")
    settingsPack.enableDht(true)
    sessionManager.applySettings(settingsPack)
    ```

2. Start `SessionManager`:
    ```kotlin
    sessionManager.start()
    ```

3. Observe logs for `LISTEN_FAILED` alerts.

---

### Expected Behavior

The `SessionManager` should bind to the specified `listenInterfaces` and report `LISTEN_SUCCEEDED` on valid ports and interfaces.

---

### Actual Behavior

- The `SessionManager` logs the following error:
  ```
  onListenFailed(): iface= , address= 0.0.0.0, port= 0, socketType= TCP, error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
  ```

---

### Environment

- **Library Version**: `jlibtorrent-1.2.19.0`
- **Platform**: Android
- **Architecture**: `arm64-v8a`
- **Device**: Tested on Samsung Galaxy M31
- **Android Version**: 12 (SDK 31)

---

### Additional Logs

Include the logs captured during the issue:
```
Applying listen interfaces: 0.0.0.0:6881-6889,[::]:6881-6889
onListenFailed(): iface= , address= 0.0.0.0, port= 0, socketType= TCP, error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
```

---

### Workaround Attempts

1. Reapplying `SettingsPack` after `SessionManager.start()` did not resolve the issue.
2. Binding to a specific port (e.g., `0.0.0.0:6881`) also failed.

---
### Here's the code for debugging purposes

[TorrentManagerJlib.txt](https://github.com/user-attachments/files/18140484/TorrentManagerJlib.txt)

**Comment by alekw** [comment]
I also confirm this on x86 android emulator, however I am not configuring any settings. I just do sessionManager.download()

**Comment by gubatron** [comment]
`6881-6889` you cannot specify a port range like that.
You can specify multiple network interfaces, each in a different port.

It's up to your client to decide on the logic of which port to pick within that range.

https://libtorrent.org/reference-Settings.html

**Comment by gubatron** [comment]
`Binding to a specific port (e.g., 0.0.0.0:6881) also failed.`

Make sure that port isn't used by another process. We cannot replicate this bug.

**Comment by LakshmanSagar** [comment]
Even if I don't allocate specific listen interface and just do like this by @alekw , i'm still getting the same issue

2024-12-16 09:22:56.347 13029-13029 System.out              com.app.animebyte                    I  jlibtorrent: SUCCESS: Loaded jlibtorrent-1.2.19.0
2024-12-16 09:22:56.376 13029-13029 System.out              com.app.animebyte                    I  Torrent session started.
2024-12-16 09:22:56.379 13029-13100 SessionManager          com.app.animebyte                    I  onListenFailed(): iface= , address= 0.0.0.0, port= 0, socketType= TCP, errorCode= com.frostwire.jlibtorrent.ErrorCode@d3ed512
2024-12-16 09:22:56.380 13029-13100 SessionManager          com.app.animebyte                    I  onListenFailed(): error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
2024-12-16 09:22:56.443 13029-13029 MSHandlerLifeCycle      com.app.animebyte                    I  removeMultiSplitHandler: no exist. decor=DecorView@b636d34[MainActivity]
2024-12-16 09:22:56.444 13029-13051 OpenGLRenderer          com.app.animebyte                    D  setSurface called with nullptr

> I also confirm this on x86 android emulator, however I am not configuring any settings. I just do sessionManager.download()

**Comment by LakshmanSagar** [comment]
- **Issue:** Using `jlibtorrent` on Android 12 results in `onListenFailed` errors when attempting to listen on the default interface (`0.0.0.0`).
- **Environment:**  
  - jlibtorrent version: 1.2.19.0  
  - Android version: 12  
  - Architecture: `arm64-v8a`

**Logs:**
```
I  jlibtorrent: SUCCESS: Loaded jlibtorrent-1.2.19.0
I  Torrent session started.
I  onListenFailed(): iface= , address= 0.0.0.0, port= 0, socketType= TCP, errorCode= com.frostwire.jlibtorrent.ErrorCode@52a7c53
I  onListenFailed(): error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
```

**Workarounds Tried:**
- Changing ports
- Modifying interfaces
- Default configurations

Issue persists regardless of the setup.

Please reopen the issue @gubatron

**Comment by proninyaroslav** [comment]
@LakshmanSagar 
This message is harmless, just libtorrent reporting the fact that for that particular interface, Android is unable bind a port. This is the expectation with the Android 12+ restrictions.

**Comment by gubatron** [comment]
absolutely.

**Comment by gubatron** [comment]
## Environment:
 - jlibtorrent version: `1.2.19.0`
 - Android version: `12` API `31`
 - Architecture: `arm64-v8a`
 
 
 It eventually binds to a port for both TCP and UDP transports
 ```
 2024-12-16 11:59:40.464  5216-5300  SessionManager          com.frostwire.android                I  onListenFailed(): error_message=listening on 0.0.0.0:0 (device: ) failed: [enum_route] [TCP] Operation not supported on transport endpoint
2024-12-16 11:59:40.464  5216-5300  BTEngine                com.frostwire.android                I  Listen failed on endpoint: 0.0.0.0:0 type:TCP (error: Operation not supported on transport endpoint/value=95)
2024-12-16 11:59:40.466  5216-5300  BTEngine                com.frostwire.android                I  Listen succeeded on endpoint: 0.0.0.0:40379 type:TCP
2024-12-16 11:59:40.472  5216-5300  BTEngine                com.frostwire.android                I  Listen succeeded on endpoint: 0.0.0.0:40379 type:UDP
```

**Comment by proninyaroslav** [comment]
> So, the app shouldwork fine,right?

Yes, I just ignore this message by default, but the user can enable it if desired.

**Comment by gubatron** [comment]
> I encountered the same issue on the Android platform. I eventually found out that the jlibtorrent interface needs to be wrapped as a service running in the background. Otherwise, once the method call finishes and the method exits, the jlibtorrent data gets destroyed.

Correct!
